src/js/controller/model.ts

```diff
@@ -1,13 +1,69 @@
 import { OS } from 'environment/environment';
 import SimpleModel from 'model/simplemodel';
 import { INITIAL_PLAYER_STATE, INITIAL_MEDIA_STATE } from 'model/player-model';
-import { STATE_IDLE } from 'events/events';
+import { InternalPlayerState, STATE_IDLE } from 'events/events';
 import { isValidNumber, isNumber } from 'utils/underscore';
 import { seconds } from 'utils/strings';
 import Providers from 'providers/providers';
+import { StreamType } from '../providers/utils/stream-type';
+import { DefaultProvider, GenericObject, PlaylistItemType, TextTrackLike } from '../types/generic.type';
+import { QualityLevel } from '../providers/data-normalizer';
+import Item from 'playlist/item';
+
+type AutoStart = boolean | 'viewable';
+
+type PlayerModelAttributes = {
+    _destroyed: boolean;
+    audioMode: boolean;
+    autostart: AutoStart;
+    autostartMuted: boolean;
+    bandwidthEstimate: number;
+    bitrateSelection: number | null;
+    captionLabel: string;
+    captionsIndex: number;
+    captionsTrack: TextTrackLike;
+    controlsEnabled: boolean;
+    defaultPlaybackRate: number;
+    dvrSeekLimit: number;
+    flashBlocked: boolean;
+    flashThrottle?: boolean;
+    fullscreen: boolean;
+    instreamMode: boolean;
+    item: number;
+    itemMeta: GenericObject;
+    itemReady: boolean;
+    mediaModel: MediaModel;
+    minDvrWindow: number;
+    mute: boolean;
+    nextUp: Item;
+    playbackRate: number;
+    playlist: PlaylistItemType[];
+    playlistItem: PlaylistItemType | null;
+    playOnViewable: boolean;
+    playRejected: boolean;
+    provider: DefaultProvider;
+    qualityLabel: string;
+    renderCaptionsNatively: boolean;
+    state: InternalPlayerState;
+    streamType: StreamType;
+    supportsPlaybackRate: boolean;
+    volume: number;
+}
+
+interface Model {
+    readonly attributes: Partial<PlayerModelAttributes>;
+    addAttributes(attributes: Partial<PlayerModelAttributes>): void;
+    get<K extends keyof PlayerModelAttributes>(attr: K): PlayerModelAttributes[K];
```

Comment:
whoa 🤯

Reply:
These docs helped:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html
https://www.typescriptlang.org/docs/handbook/utility-types.html

Reply:
Also the type checking for this interface only works by naming the interface after the class. When I tried `class Model extends SimpleModel implements ModelInterface` the interface of `SimpleModel` took precedence.

Reply:
awesome!